WIP: Implement AST style parsing.

twavv opened the pull request:
Credit to @aaronsheldon for most of this. This is the very beginning.

Ping @shashi @piever (I can't help rewriting things, I'm sorry).

twavv committed:
Update how observable values are updated/fetched to match new WebIO.

twavv committed:
Start implementing AST style parsing.

piever commented on src/ast.jl:
```diff
@@ -0,0 +1,44 @@
+    #     end
+    # end
+
+    JSAST(h) = new(h, [])
```
This shouldn't be internal constructors: I think the struct can be defined without internal constructors and you can add (outside of the struct definition) some external constructors (e.g `JSAST(h::Symbol, body::Tuple) = JSAST(h, collect(body))`).

piever commented on src/JSExpr.jl:
```diff
@@ -1,207 +1,91 @@
+function crawl(ex::Expr)::JSNode
+    # Recurse into expressions
+    if hasmethod(crawl, typeof((Val(ex.head), ex.args...)))
+        return crawl(Val(ex.head), ex.args...)::JSNode
```
Given that the function has an explicit return type, you probably don't need one here.

piever commented on src/JSExpr.jl:
```diff
@@ -1,207 +1,91 @@
+"""
+function crawl(ex::Expr)::JSNode
+    # Recurse into expressions
+    if hasmethod(crawl, typeof((Val(ex.head), ex.args...)))
```
```julia
if hasmethod(crawl, ...)
   crawl(...)
else
    ...
end
```
seems a bit strange. Couldn't you define a fallback:
```julia
function crawl(::Val{sym}, args...) where sym
     ex = Expr(sym, args...)
     error("Expression $ex not supported.")
end
```

and just call `crawl(Val(ex.head), ex.args...)` normally? If there is a method it will get it, otherwise you hit the fallback and error.

piever replied:
If the `hasmethod` approach turns out to be necessary, rather than `hasmethod(f, typeof(args))` you can just do `applicable(f, args....)`.

twavv commented:
The biggest roadblock for me right now is how to implement interpolation...

piever commented on src/JSExpr.jl:
```diff
@@ -1,207 +1,91 @@
-    end
+    # Push terminals, when all else fails try interpolation. This will be a source of bugs
+    # because there are no guarantees the string representation will be sensible.
+    elseif hasmethod(string, Tuple{T})
```
This is always true. As above, it can probably be simplified by adding a fallback ``JSSymbol(x::Any) = JSSymbol(JSString(string(ex)))` and calling `JSSymbol(ex)` directly here rather than the `if hasmethod` approach.

piever commented on src/call.jl:
```diff
@@ -0,0 +1,12 @@
+
+    # Otherwise pass as a call
+    else
+        return JSAST(:call, [crawl(m), crawl.(b)...])
```
Again, probably easier to add a `crawl(::Val{:call}, ::Val{m}, b...) where {m} = JSAST(:call, [crawl(m), crawl.(b)...])` fallback.

piever commented on src/literals.jl:
```diff
@@ -0,0 +1,16 @@
+    if hasmethod(crawl, (Val{:symbol}, Val{sym}))
+        return crawl(Val(:symbol), Val(sym))
+    end
+    return JSSymbol(sym)
```
This also could be a fallback `crawl(::Val{:symbol}, ::Val{sym}) where {sym} = JSSymbol(sym)`

piever commented on src/ast.jl:
```diff
@@ -0,0 +1,44 @@
+A JavaScript symbol. Our definition of symbol is a bit broader than Julia's
+definition since we also include all literals.
+
+TODO: Rename to `JSTerminal`?
```
:+1:

piever commented:
> The biggest roadblock for me right now is how to implement interpolation...

IIUC the strategy is to convert to some sort of AST at macro expand time (when the value of interpolated things is not accessible) and then turn it directly into a string with `deparse`. I think this strategy needs to change a bit to work with interpolation.

You need to:

- use `esc(....)` in the right place to make sure you are in the right context at the end to evaluate the interpolated things (this is always a bit tricky)
- Handle `Expr(:$, sym)` by putting some placeholder in the JSAST and at the end rather than `Expr(:call, :jsstring, deparse(crawl(ex)))` you should return something like `Expr(:call, :replace_interpolated, deparsed_string, vars...)` where `deparse_str` is the string with the placeholders for the interpolated variables and `vars` are those variables (which will be evaluated as they are part of the expression returned by the macro). 

Hope this helps!

twavv committed:
Make crawl return Exprs, implement interpolation.

twavv committed:
Infix operators.

twavv committed:
Merge branch 'master' into ast

twavv committed:
Simply default crawl.

twavv committed:
Merge branch 'ast' of github.com:travigd/JSExpr.jl into ast

twavv committed:
More test cases.

twavv committed:
Add functions.

shashi commented:
I'd just like to voice my initial opinion here:

This is great @aaronsheldon, @travigd and @piever !

twavv committed:
Change tuple handling to arrays.

twavv committed:
Implement if parsing, rework some things about functions.

twavv committed:
Implement while loops.

twavv committed:
Add for loops.

twavv committed:
Refactor some details around crawling function calls.

twavv committed:
Add convenience translation for range.

twavv committed:
Add ref and special case WebIO and Observables.

twavv committed:
Add standard indexing.

twavv committed:
Implement updating WebIO Observables.

twavv committed:
Change tuple handling.

twavv committed:
Modify implicit return statement behavior.

twavv committed:
Add more infix operators.

twavv committed:
Add object construction syntaxes.

twavv committed:
Fix Symbol and QuoteNode handling.

twavv commented:
@piever @pfitzseb @shashi @SimonDanisch @andreasnoack @rdeits  (just naming whoever I can think of for now - sorry if you're not interested or I missed you!)

I have the formalized version working. It's currently incompatible with Blink because of some things that it does that use JSExpr's internal-ish API but works with Interact and Jupyter. For @rdeits, don't worry, I checked MeshCat and it works with all of the examples in the readme. I'd be willing to fix Blink once it's released.

If there are no objections, I'd like to fix the merge conflicts, change up the WebIO dependency, and release this version as v1.0 (this version should be compatible with anyone currently using the public API, but release a new major version is probably the route to go here).

By change up the WebIO dependency, I mean to reverse the direction of the dependency so that WebIO hooks into JSExpr. This is much easier with this version. Essentially, it would entail moving the JSString type out of WebIO and into JSExpr and moving the stuff in `webio.jl` into WebIO.

twavv commented:
There are actually a few (maybe only one?) things I think I'd like to discuss changing before v1.0 that actually would be breaking.

Specially parsing `Dict(...)` inside of JSExpr doesn't make me super happy because that could be a valid JavaScript function name and we don't transform any other functions* (such as `print`). I added a braces syntax to this PR; I'm thinking it might be polarizing but if it's okay I think it might be okay to use that instead of adding special processing for `Dict` calls.
```julia
# Braces syntax
@js myDict = {
    :foo => "bar",
    "spam" => "eggs",
}
```

*In this PR, I actually did special case the `:` method (range constructor) for convenience because it's not a valid identifier in JS:
```julia
@js for i in 1:20
    console.log(i)
end
```
The resulting code is not exactly pretty, but it's the best we can do™ without much more effort:
```js
for (let i of (new Array(20).fill(undefined).map((_, i) => i + 1))) { console.log(i); }
```

**Edit:** I changed for loops to use the `for (... of ...)` syntax which works better with arrays, but also doesn't work with vanilla objects anymore.
```js
> for (const x in myArr) { console.log(x); }
0
1
> for (const x of myArr) { console.log(x); }
foo
bar
> const myObj = { foo: "bar", spam: "eggs" };
> for (const key in myObj) {
...   console.log(key);
... }
foo
spam
> for (const key of myObj) {
...   console.log(key);
... }
Thrown:
TypeError: myObj is not iterable
```

The reason for the difference is that `in` iterates over all the keys of an object in JS (and arrays are, conceptually, just JS objects whose keys are `1`, `2`, ...). I'm not sure if iterating over objects was a common thing to do using JSExpr.

twavv committed:
Fix parenthesization/associativity of infix operators.

twavv committed:
Merge branch 'master' into ast

twavv committed:
Remove unnecessary syntax.jl file.

twavv committed:
Modify macro handling.

twavv commented:
Continuing in #31.

rdeits commented:
Hey, sorry I didn't respond at the time, but I just wanted to chime in to say that this seems like a really good idea. Thanks for checkout out MeshCat, too! I think the special syntax for dictionaries makes sense, and not special-casing `dict()` seems like a good idea.
